client_plugin/drivers/vfile/kvstore/etcdops/etcdops.go

```diff
@@ -289,20 +305,100 @@ func (e *EtcdKVS) joinEtcdCluster(leaderAddr string) error {
 	}
 
 	// start the routine for joining an etcd cluster
-	go etcdService(lines)
+	e.etcdStartService(lines)
 
 	// check if successfully joined the etcd cluster, then start the watcher
 	return e.checkLocalEtcd()
 }
 
-// etcdService function starts a routine of etcd
-func etcdService(cmd []string) {
-	_, err := exec.Command("/bin/etcd", cmd...).Output()
+// leaveEtcdCluster function is called when a manager is demoted
+func (e *EtcdKVS) leaveEtcdCluster() error {
+	nodeAddr := e.nodeAddr
+	nodeID := e.nodeID
+
+	etcd, err := addrToEtcdClient(nodeAddr)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"nodeAddr": nodeAddr,
+				"nodeID": nodeID},
+		).Error("Failed to create ETCD client from own address")
+		return err
+	}
+	defer etcd.Close()
+
+	// list all current ETCD members
+	ctx, cancel := context.WithTimeout(context.Background(), etcdRequestTimeout)
+	lresp, err := etcd.MemberList(ctx)
+	cancel()
+	if err != nil {
+		log.WithFields(
+			log.Fields{"nodeAddr": nodeAddr,
+				"error": err},
+		).Error("Failed to list member for ETCD")
+		return err
+	}
+
+	peerAddr := etcdScheme + nodeAddr + etcdPeerPort
+	for _, member := range lresp.Members {
+		// loop all current etcd members to find if there is already a member with the same peerAddr
+		if member.PeerURLs[0] == peerAddr {
+			log.WithFields(
+				log.Fields{"nodeID": nodeID,
+					"peerAddr": peerAddr},
+			).Info("Remove self from ETCD member due to demotion. ")
+
+			ctx, cancel = context.WithTimeout(context.Background(), etcdRequestTimeout)
+			_, err = etcd.MemberRemove(ctx, member.ID)
+			cancel()
+			if err != nil {
+				log.WithFields(
+					log.Fields{"peerAddr": peerAddr,
+						"member.ID": member.ID},
+				).Error("Failed to remove member for ETCD ")
+				return err
+			}
+			// the same peerAddr can only join at once. no need to continue.
+			break
+		}
+	}
+
+	e.etcdStopService()
+	return nil
+}
+
+// etcdStartService function starts an ETCD process
+func (e *EtcdKVS) etcdStartService(lines []string) {
+	cmd := exec.Command("/bin/etcd", lines...)
```

Comment:
We should use systemd to manage services. Running daemons ourselves means we are in charge of resource allocation and restart on issues....  If we do no have a tracking issue for this, please do open one

Reply:
Agree.
Issue created: https://github.com/vmware/docker-volume-vsphere/issues/1873